### DISCUSSION TITLE
HMR: Websocket is closed before the connection is established

### DISCUSSION THREAD

**Discussion by krish-penumarty** [vercel/next.js#40533] [Help]
Hello all! I could really use some help after we recently upgraded to Next.js v12. We started seeing errors `Websocket connection to 'wss://localhost:8443/assets-marketplace/_next/webpack-hmr' failed: WebSocket is closed before the connection is established`. 

We are using `https` for our localhost and thus based on the [source code](https://github.com/genetschneider/next.js-1/blob/e4085069e4a54373d765d1aeca5e71f80020e22b/packages/next/client/dev/error-overlay/websocket.ts#L35) it uses `wss`. I followed the [upgrade guide ](https://nextjs.org/docs/upgrading#nextjs-hmr-connection-now-uses-a-websocket) and added the location block to support WebSocket. However, we are still seeing this issue. Wondering if there are any tips on how to handle this?

This is the error I see on our Nginx logs
```
"GET /assets-marketplace/favicon.ico HTTP/1.1" 200 4286 "https://localhost:8443/marketplace" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36"
2022/09/14 00:16:52 [info] 38#38: *33 epoll_wait() reported that client prematurely closed connection, so upstream connection is closed too while sending request to upstream, client: 172.17.0.1, server: *.localhost, request: "GET /assets-marketplace/_next/webpack-hmr HTTP/1.1", upstream: "http://<IP_ADDRESS>:<PORT>/assets-marketplace/_next/webpack-hmr", host: "localhost:8443"
```

Running v12.2.5

**Comment by krish-penumarty** [comment]
Bump! Wondering if the Next.js team could provide some guidance here?

**Comment by kasber-developer** [comment]
I have same error on Next 13. Any solution?

**Comment by igl** [comment]
Same here. First i thought it was our nginx proxy, but it also happens when opening nextjs directly.

> WebSocket connection to 'ws://localhost:9000/_next/webpack-hmr' failed: WebSocket is closed before the connection is established.
handleDisconnect @ websocket.js?a234:45
eval @ websocket.js?a234:50

**Reply by igl** [reply]
It's the `assetPrefix` option. Make sure to also add a route for `webpack-hmr` separately from your assetPrefix route:

<img width="432" alt="image" src="https://user-images.githubusercontent.com/293519/204761413-699ef5d4-daf0-4848-ab36-35030fe5fd90.png">

**Reply by krish-penumarty** [reply]
We tried that and it still didn't work for us :( I'll keep digging and follow back up here.

**Comment by cherealnice** [comment]
Any luck with this? I'm running into the same issue

**Comment by joao-pedro-alves** [comment]
For those facing the same issue, I could fix it by using this nginx configuration:

```
location /_next/webpack-hmr {
        proxy_pass http://workspace:3000/_next/webpack-hmr;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
}
```

Here is the full **server** section (i'm using docker btw)
```
server {
    listen 80;
    listen [::]:80;
    server_name app.cyberclass.local;

    location / {
        proxy_pass  http://workspace:3000/;
        proxy_set_header    Host                $host;
        proxy_set_header    X-Real-IP           $remote_addr;
        proxy_set_header    X-Forwarded-Host    $host;
        proxy_set_header    X-Forwarded-Server  $host;
        proxy_set_header    X-Forwarded-Proto   $scheme;
        proxy_set_header    X-Forwarded-For     $remote_addr;
        proxy_redirect off;
        proxy_connect_timeout 90s;
        proxy_read_timeout 90s;
        proxy_send_timeout 90s;
    }

    location /_next/webpack-hmr {
        proxy_pass http://workspace:3000/_next/webpack-hmr;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
    }
}
```

**Comment by rimzzlabs** [comment]
Im having the same Issue

im using windows 10, no nginx, this error only shows on dev mode, anyone can explain why?
I've searched the internet and I only come up with nginx config
